### PR TITLE
[NVIDIA XLA]Replace broadcast of trivial matrix bias by optional custom-call for CUDA >= 12

### PR DESCRIPTION
**Pull request by wenscarl**
Replace broadcast of trivial matrix bias by a scalar under CUDA > 12. Nearly all F8 unittests will get impacted but only 1 unittest is included at the moment for the purpose of demonstration. 

Broadcasting trivial matrix bias takes up about 10% time of the actually matmul. 
image

### CODE REVIEW

**Comment by pjannaty** [comment]
@reedwm this seems to be of high priority given @wenscarl is reporting the perf hit is ~ 10% of the actual matmul call.

**Comment by pjannaty** [comment]
Seems like all the unit tests will have to change to accommodate this change. Wenscarl has only implemented a single unit test to show the shape of things to come and is seeking feedback whether the new paradigm works. @philipphack will also give feedback.

**Review by philipphack — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:886]: Why create a `c` at all, if it's just going to be replaced by an empty buffer in the thunk? We don't have to pass `c` to the custom call.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:879]: I saw [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc;l=728?q=12000&ss=tensorflow%2Ftensorflow) is > 12000. So follow that.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:839]: Yes. Isolate `pad_shape` to be used in determining if c is padded.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:883]: Yes, but c has to be something, in this case I made it a scalar, to get passed into `HloInstruction::CreateCustomCall()`.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:886]: Yes, but c has to be something, in this case I made it a scalar, to get passed into HloInstruction::CreateCustomCall().

Comment (philipphack) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:883]: We could create the Custom Call without c when beta is zero and CUDA is version 12 or higher.

Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:883]: Yes, it's possible, but that requires lots of plumbing work and hence generating quite several compile conditional flags. A scalar placeholder operand is used here for convenience.

Comment (reedwm) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:883]: I would much prefer creating the custom call without C when beta is zero. Otherwise it's inconsistent with the non-FP8 case and it's confusing how the custom call has an unused placeholder for C that is unused.

Having a lot of `CUDA_VERSION >= 12000` checks would be annoying. Is it possible to set beta to 1 for CUDA 11 when creating a zero bias, in which case we wouldn't need such checks and instead could assume we have a matrix bias iff beta is nonzero. Alternatively, perhaps we should only support this pass for CUDA 12 and above.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Comment by philipphack** [comment]
Version 11 doesn't support non-zero beta. I think that's why we added the version guard [here](https://github.com/tensorflow/tensorflow/blame/0e51c74cd0fa8bef5d9cdbea76f3387a1baae01b/tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc#L728).

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:883]: If to implementing creating custom call without C to make it consistent with the non-FP8 , the operands[2] can be matrix or vector bias, check [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/translate/mhlo_to_lhlo_with_xla/mhlo_to_lhlo_with_xla.cc;l=971), which is what non-FP8 case does. But this plan was rejected in that PR. Check earlier discussion [here](https://github.com/tensorflow/tensorflow/pull/60020#discussion_r1145622210).

**Comment by wenscarl** [comment]
Changes are made with CUDA_VERSION >= 12000 guards removed.

**Comment by reedwm** [comment]
Can you sync with the master branch to get changes from https://github.com/tensorflow/tensorflow/pull/60502?

**Comment by wenscarl** [comment]
> Can you sync with the master branch to get changes from #60502?

Rebased to current master.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:729]: Revert whitespace changes

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:862]: Use `pad_shape` it in `pad_operand` to avoid the duplicated code which calls `RoundUpTo`.

Also it might be worth making `pad_shape` and `pad_operand` top-level functions instead of lambda functions.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:1172]: The non-FP8 case doesn't have this logic. It unconditionally gets C, which IIUC is simply the output if there is no matrix bias. For consistency with the non-FP8 case, can we also uncondtionally get C here? This would also simply some logic in cublas_lt_matmul_think.cc and matmul_utils.h.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — COMMENTED** [review]
Since @wenscarl is currently OOO, I'm approving and will address all comments when merging

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Comment by tdanyluk** [comment]
As part of the OpenXLA PR rotation:
I'll re-trigger the presubmits as soon as I get permissions.

**Comment by reedwm** [comment]
Since the PR author is OOO, I was going to fix some issues when merging. But it's too hard to test on H100s when merging, so instead I'll commit directly to this PR's branch. So let's not merge this PR yet.

**Comment by reedwm** [comment]
@philipphack in the FP8 case, when there is no matrix bias, this PR currently sets C to nullptr. But in the FP16 case, when there is no matrix bias, instead C is set to D

I would prefer that the FP16 and FP8 cases be unified. Is there a fundamental reason why it's important for C to be nullptr in the FP8 case but not in the non-FP8 case? If there is no difference in cublas LT between FP8 and FP16 w.r.t. whether C should be nullptr or equal to D, I will set C to D in this PR. (Since @wenscarl is OOO I am addressing my own comments to this PR before merging it)

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Pull request merged**
